torch/csrc/jit/testing/file_check.cpp

```diff
@@ -133,32 +156,124 @@ struct FileCheckImpl {
       groups.push_back({check});
     } else {
       auto& last_group = groups.back();
-      if (last_group.at(0).type_ == type) {
+      if (last_group.at(0).type_ == check.type_) {
         last_group.push_back(check);
       } else {
         groups.push_back({check});
       }
     }
-
     has_run = false;
   }
 
+  TORCH_API void addCheck(
+      CheckType type,
+      const std::string& s,
+      c10::optional<size_t> count = c10::nullopt) {
+    addCheck(Check(type, s, std::move(count)));
+  }
+
   bool has_run = false;
 
   friend std::ostream& operator<<(std::ostream& out, const FileCheckImpl& fc);
 
  private:
+  void parseStrings(std::shared_ptr<std::string>& checks_file) {
+    std::vector<Check> operands;
+    const static std::vector<std::pair<CheckType, std::string>> check_pairs = {
+        {CHECK, ": "},
+        {CHECK_NEXT, "-NEXT: "},
+        {CHECK_SAME, "-SAME: "},
+        {CHECK_NOT, "-NOT: "},
+        {CHECK_DAG, "-DAG: "},
+        {CHECK_COUNT, "-COUNT-"}, // needs special parsing
+    };
+    const std::string prefix = "# CHECK";
+
+    size_t start = 0;
+    size_t num_checks = 0;
+    start = checks_file->find(prefix, start);
+
+    while (start != std::string::npos) {
+      bool found_match = false;
+      for (const auto& check_pair : check_pairs) {
+        const std::string& check_suffix = check_pair.second;
+        auto suffix_pos = checks_file->find(check_suffix, start);
+        if (suffix_pos != start + prefix.size()) {
+          continue;
+        }
+        size_t end_check_string = suffix_pos + check_suffix.size();
+        CheckType type = check_pair.first;
+        c10::optional<size_t> count = c10::nullopt;
+        auto end_line = checks_file->find("\n", end_check_string);
+        bool exactly = false;
+        if (type == CHECK_COUNT) {
+          const std::string exact = "EXACTLY-";
+          if (checks_file->find(exact, end_check_string) == end_check_string) {
+            exactly = true;
+            end_check_string += exact.size();
+          }
+          size_t end = assertFind(
+              SourceRange(checks_file, end_check_string, end_line), ":");
+          count = std::stoll(
+              checks_file->substr(end_check_string, end - end_check_string));
+          end_check_string = end + 2; // add ':' and the space
+        }
+        auto check = Check(
+            type,
+            checks_file->substr(end_check_string, end_line - end_check_string),
+            count);
+        addCheck(check);
+        if (exactly) {
+          addCheck(CHECK_NOT, check.search_str_);
+        }
+        start = end_line;
+        found_match = true;
+        break;
+      }
+      if (!found_match) {
+        std::ostringstream ss;
+        ss << "Could not parse check at:\n";
+        SourceRange(checks_file, start, start + prefix.size()).highlight(ss);
+        ss << "Check for bad input.";
+        has_run = true;
+        throw std::runtime_error(ss.str());
```

Comment:
I'm not sure we should throw every time we weren't able to parse a check tag. I can easily imagine having an input with comments "# CHECK THAT SOMETHING IS CORRECT" or something like that - we shouldn't throw an exception when we see that. I do agree that it might be helpful to throw in cases like "# CHECK-COUNT: qqq" though, maybe we should just be a bit more selective when we do that.

Reply:
I would rather error on the side of false positive than false negatives, although I could see writing something like # CHECK ... and having an error message be annoying.

Reply:
One other thing that might be too restrictive is that we require the there is exactly one space between '#' and 'CHECK' - maybe we could loosen that a bit too.